storage/cloudspanner/tree_storage.go

```diff
@@ -0,0 +1,523 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudspanner
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/google/trillian"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/cache"
+	"github.com/google/trillian/storage/cloudspanner/spannerpb"
+	"github.com/google/trillian/storage/storagepb"
+	"github.com/google/trillian/trees"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	// ErrNotFound is returned when a read/lookup fails because there was no such
+	// item.
+	ErrNotFound = status.Errorf(codes.NotFound, "not found")
+
+	// ErrNotImplemented is returned by any interface methods which have not been
+	// implemented yet.
+	ErrNotImplemented = errors.New("not implemented")
+
+	// ErrTransactionClosed is returned by interface methods when an operation is
+	// attempted on a transaction whose Commit or Rollback methods have
+	// previously been called.
+	ErrTransactionClosed = errors.New("transaction is closed")
+
+	// ErrWrongTXType is returned when, somehow, a write operation is attempted
+	// with a read-only transaction.  This should not even be possible.
+	ErrWrongTXType = errors.New("mutating method called on read-only transaction")
+
+	// errFinished is only used to terminate spanner reads early, once all
+	// required data has been read. It should never be returned to a caller.
+	errFinished = errors.New("read complete")
+)
+
+const (
+	subtreeTbl   = "SubtreeData"
+	colBucket    = "Bucket"
+	colSubtree   = "Subtree"
+	colSubtreeID = "SubtreeID"
+	colTreeID    = "TreeID"
+	colRevision  = "Revision"
+)
+
+// treeStorage provides a shared base for the concrete Spanner-backed
+// implementation of the Trillian storage.LogStorage and storage.MapStorage
+// interfaces.
+type treeStorage struct {
+	admin  storage.AdminStorage
+	opts   TreeStorageOptions
+	client *spanner.Client
+}
+
+// TreeStorageOptions holds various levers for configuring the tree storage instance.
+type TreeStorageOptions struct {
+	// ReadOnlyStaleness controls how far in the past a read-only snapshot
+	// transaction will read.
+	// This is intended to allow Spanner to use local replicas for read requests
+	// to help with performance.
```

Comment:
Can we give a hint on useful range of values here? Typically a few minutes or something maybe.

Reply:
I don't think we know at the moment, and this could also change with time...  this sort of info would probably be better in an operations doc (where it might just be a link to the CS docs?)

Reply:
True. We could maybe say what happens if you increase it though.

Reply:
Isn't that covered by the first sentence?
Maybe I'm missing something though; do you have a concrete suggestion for how to reword/extend it?

Reply:
I just wanted to avoid people thinking that pulling this lever increases performance. Could maybe link to Spanner doc?

Reply:
oke doke, done